src/main/java/org/opensearch/neuralsearch/search/query/HybridCollectorManager.java

```diff
@@ -118,6 +119,27 @@ public Collector newCollector() {
      */
     @Override
     public ReduceableSearchResult reduce(Collection<Collector> collectors) {
+        final List<HybridTopScoreDocCollector> hybridTopScoreDocCollectors = getHybridScoreDocCollectors(collectors);
+        if (hybridTopScoreDocCollectors.isEmpty()) {
+            throw new IllegalStateException("cannot collect results of hybrid search query, there are no proper score collectors");
+        }
+
+        List<ReduceableSearchResult> results = new ArrayList<>();
+        DocValueFormat[] docValueFormats = getSortValueFormats(sortAndFormats);
+        for (HybridTopScoreDocCollector hybridTopScoreDocCollector : hybridTopScoreDocCollectors) {
+            List<TopDocs> topDocs = hybridTopScoreDocCollector.topDocs();
+            TopDocs newTopDocs = getNewTopDocs(
+                getTotalHits(this.trackTotalHitsUpTo, topDocs, hybridTopScoreDocCollector.getTotalHits()),
+                topDocs
+            );
+            TopDocsAndMaxScore topDocsAndMaxScore = new TopDocsAndMaxScore(newTopDocs, hybridTopScoreDocCollector.getMaxScore());
+
+            results.add((QuerySearchResult result) -> reduceCollectorResults(result, topDocsAndMaxScore, docValueFormats, newTopDocs));
+        }
+        return reduceSearchResults(results);
```

Comment:
So what I can see here is first for every collector results we are converting the to TopDocsAndMaxScore(the format what hybrid query needs) and then merging all of them.

Can we do this reverse, where we first merge the results and then we convert the results to TopDocsAndMaxScore. I feel in this way we can avoid a lot of duplicate iterations.

thoughts? @martin-gaievski , @vibrantvarun

Reply:
I had same idea initially, but unfortunately it's not feasible. Main reason is that every run of reduce method doesn't know about other reduces. We need to put collected results back to QueryResults object, so if we had such knowledge we can merge individual results and then update QueryResult at the end. But because such knowledge is missing and reduce is called b core (and differently depending on w/o concurrent segment search, and maybe also that different slicing mechanism that you've shared) then our reduce call should be stateless. 
I'm running benchmarks now to check if there is impact of performance